### ISSUE TITLE
Maximum Call Stack Exceeded using Property Fields

### ISSUE THREAD

**Issue by AndiLi99** [amcharts/amcharts4#946]
Setting colour using propertyFields creates many segments, which exceeds the call stack limit.

Here is a codepen: https://codepen.io/andili99/pen/pGpweW

Is there anything I can do to avoid this issue while using custom colours for each segment, and having a high number of segments (6000+)?

**Comment by AndiLi99** [comment]
Here's the full stack trace I get. I'm using the latest version through the CDN (4.1.5)

```
RangeError: Maximum call stack size exceeded
    at new e (LineSeriesSegment.ts:101)
    at e.t.clone (core.js:formatted:18019)
    at e.create (core.js:formatted:37803)
    at t._create (LineSeries.ts:227)
    at t.returnItem (core.js:formatted:23699)
    at t.getFirst (core.js:formatted:23671)
    at e.openSegment (LineSeries.ts:444)
    at e.closeSegment (LineSeries.ts:564)
    at e.openSegment (LineSeries.ts:500)
    at e.closeSegment (LineSeries.ts:564)
    at e.openSegment (LineSeries.ts:500)
    ...
    repeat 4000 times
    ...
    at e.closeSegment (LineSeries.ts:564)
    at e.openSegment (LineSeries.ts:500)
    at e.closeSegment (LineSeries.ts:564)
    at e.openSegment (LineSeries.ts:500)
    at e.closeSegment (LineSeries.ts:564)
    at e.openSegment (LineSeries.ts:500)
    at e.validate (LineSeries.ts:361)
    at core.js:formatted:17270
    at core.js:formatted:16854
    at u (core.js:formatted:16906)
    at Object.Qkdp.e.each (core.js:formatted:16853)
    at t.update (core.js:formatted:17253)
    at core.js:formatted:17334
```

**Comment by martynasma** [comment]
That does not seem viable. Each color change creates a new masked "copy" of the series, resulting in a lot of elements.

Besides, how exactly user is supposed to perceive 6000 color stripes on a simple chart? :)

**Comment by AndiLi99** [comment]
I dug into the source code and it seems the issue stems from mutual recursion. openSegment() calls closeSegment() which in turn calls openSegment(), and repeat. Normally this would be okay since it's a tail call and would be tail call optimised, but this is javascript and it's not supported.

I mocked up an iterative solution in the typescript file for LineSeries, but I can't seem to get it to run. I have a question, do the typescript files get compiled into javascript or is the source typescript file read directly?

**Comment by martynasma** [comment]
Hey, just wanted to update to you that we are reviewing the changes from PR.

Will let you know when this is in place.

**Comment by martynasma** [comment]
This has now been merged in internal repo. Will release in the next few days.

Thanks!

**Comment by martynasma** [comment]
Fixed in [4.1.8](https://github.com/amcharts/amcharts4/releases/tag/4.1.8).

## [4.1.8] - 2019-02-15

### Added
- JSON: You can now specify both list of items and template settings for `ListTemplate`, e.g.: `titles: { template: { ... }, values: [{ ... }] }`. Specifying it the old way will still work.

### Changed
- `DateAxis` will now pay attention to `dateFormatter.firstDayOfWeek` when grid is in "weekly" mode.

### Fixed
- Using `proprtyFields` on a `LineSeries` sometimes could result in an error. (fix by @AndiLi99)
- French locale (fr_FR) updated for correct decimal/thousands separators.
- Export: `scale` in image export options was being ignored.
- JSON: `cursor.snapToSeries` was not working in JSON configs.
- Selection was acting funky when zooming `DateAxis` and cursor moved out of plot area.
- An infinite error message loop fixed which was happening if series was added and there was no X or Y axes defined.
- Tooltip on the last available data item was left visible even if cursor moved to the date where the series had no data points.
- When `snapToSeries` was enabled in `XYCursor`, vertical line was not shown.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.

**Comment by AndiLi99** [comment]
Great! Thanks for including it.
